src/parsers/espresso/espresso.js

```diff
@@ -0,0 +1,389 @@
+import { ATOMIC_COORD_UNITS, coefficients } from "@exabyte-io/code.js/dist/constants";
+import _ from "underscore";
+import s from "underscore.string";
+
+import { ConstrainedBasis } from "../../basis/constrained_basis";
+import { primitiveCell } from "../../cell/primitive_cell";
+import { Lattice } from "../../lattice/lattice";
+import { LATTICE_TYPE } from "../../lattice/types";
+import math from "../../math";
+import { parseFortranFile } from "../../utils/parsers/fortran";
+import xyz from "../xyz";
+import { regex } from "./settings";
+
+/**
+ * Construct textual representation of a materialOrConfig according to Quantum ESPRESSO pw.x input format.
+ * @param {Material|Object} materialOrConfig - material class instance or its config object
+ * @return {String}
+ */
+function toEspressoFormat(materialOrConfig) {
+    const l = new Lattice(materialOrConfig.lattice);
+    const vectors = l.vectorArrays;
+    const vectorsAsString = _.map(vectors, (v) => {
+        return `${s.sprintf("%14.9f", v[0])}\t${s.sprintf("%14.9f", v[1])}\t${s.sprintf(
+            "%14.9f",
+            v[2],
+        )}`;
+    }).join("\n");
+    return s.sprintf(
+        "CELL_PARAMETERS (angstroms)\n%s\n\nATOMIC_POSITIONS (crystal)\n%s",
+        vectorsAsString,
+        xyz.fromMaterial(materialOrConfig),
+    );
+}
+
+/**
+ * @summary checks if the given fileContent is in the format of a Quantum ESPRESSO .in file.
+ * @param {String} fileContent
+ * @return {Boolean}
+ */
+function isEspressoFormat(fileContent) {
+    const espressoRegex = regex.espressoFingerprint;
+    return espressoRegex.test(fileContent);
+}
+
+/**
+ * @summary Return unit cell parameters from CELL_PARAMETERS card
+ * @param {String} text - cards data
+ * @return {{vectors: Number[][], units: String}}
+ */
+function getCellConfig(text) {
+    let vectors = [];
+    let units;
+    const cellParameters = text.match(regex.cellParameters);
+
+    if (cellParameters) {
+        // eslint-disable-next-line prefer-destructuring
+        units = cellParameters[1];
+        vectors = [
+            [
+                parseFloat(cellParameters[2]),
+                parseFloat(cellParameters[3]),
+                parseFloat(cellParameters[4]),
+            ],
+            [
+                parseFloat(cellParameters[5]),
+                parseFloat(cellParameters[6]),
+                parseFloat(cellParameters[7]),
+            ],
+            [
+                parseFloat(cellParameters[8]),
+                parseFloat(cellParameters[9]),
+                parseFloat(cellParameters[10]),
+            ],
```

Comment:
All of the above on lines 59-72 can be done via regex group matching

Reply:
It's a matter of preference. 
I'd do it with named capturing groups like this:
```
    if (match) {
        const vectors = [
            [parseFloat(match.groups.x1), parseFloat(match.groups.y1), parseFloat(match.groups.z1)],
            [parseFloat(match.groups.x2), parseFloat(match.groups.y2), parseFloat(match.groups.z2)],
            [parseFloat(match.groups.x3), parseFloat(match.groups.y3), parseFloat(match.groups.z3)],
        ];
```

While there's example of reverse function which is one liner basically, but personally I find it harder to understand correctly from the first scanning look:
```
const vectorsAsString = _.map(vectors, (v) => {
        return `${s.sprintf("%14.9f", v[0])}\t${s.sprintf("%14.9f", v[1])}\t${s.sprintf(
            "%14.9f",
            v[2],
        )}`;
    }).join("\n");
```

Although, if map functions like this are coding equivalent of Einstein summation, I'll use them